Implemented servlet 6.1 redirect with content

gregw opened the pull request:
Added option for server to generate a short html redirect body content, as per RFC9110 (default false) Allowed an aggregated servlet response content to be used if clear is false.

gregw committed:
Implemented servlet 6.1 redirect with content

Added option for server to generate a short html redirect body content, as per RFC9110 (default false)
Allowed an aggregated servlet response content to be used if clear is false.

janbartel reviewed (changes requested):
Where's the implementation of this line of the javadoc for the new `sendRedirect` method:

> This method has no effect if called from an include.

gregw committed:
Redirect is a noop in include

joakime commented on jetty-ee11/jetty-ee11-servlet/src/main/java/org/eclipse/jetty/ee11/servlet/Dispatcher.java:
```diff
@@ -664,6 +664,24 @@ public void sendRedirect(String location) throws IOException
+        @Override
+        public void sendRedirect(String location, boolean clearBuffer) throws IOException
+        {
+            // NOOP for include.
```
Would there be any benefit to have these be LOG.debug() statements that indicate the skip?  That way someone that is boggled by this change can see the output in the debug logs? vs having the source code of Jetty to know what is going on?

gregw replied:
This is not really a change in behaviour, at least not against jetty-9/10/11 as they all noop methods on include.  So we are really just extending to a few more methods and fixing cross context.  Anybody who is boggled should read the javadoc of sendRedirect and it clearly says that it is a noop for include.
So I'd rather avoid the noise of lots of debugs.

gregw committed:
Fixed init order

This style of extensibility (calling virtuals from constructors) is very fragile.

janbartel commented on jetty-ee11/jetty-ee11-servlet/src/main/java/org/eclipse/jetty/ee11/servlet/ServletApiResponse.java:
```diff
@@ -632,4 +656,132 @@ public String toString()
+        @Override
+        public void setCharacterEncoding(String charset)
+        {
+            // NOOP for include.
```
The javadoc for this method doesn't mention it isn't valid for an include. Are you sure about this?

gregw replied:
See the link to RequestDispatcher#include added above.   An include cannot modify response headers, so all the methods that can do that are noops.

janbartel commented on jetty-ee11/jetty-ee11-servlet/src/main/java/org/eclipse/jetty/ee11/servlet/ServletApiResponse.java:
```diff
@@ -632,4 +656,132 @@ public String toString()
+        @Override
+        public void setLocale(Locale loc)
+        {
+            // NOOP for include.
```
See above

janbartel commented on jetty-ee11/jetty-ee11-servlet/src/main/java/org/eclipse/jetty/ee11/servlet/ServletApiResponse.java:
```diff
@@ -632,4 +656,132 @@ public String toString()
+        @Override
+        public void setContentLength(int len)
+        {
+            // NOOP for include.
```
See above.

janbartel commented on jetty-ee11/jetty-ee11-servlet/src/main/java/org/eclipse/jetty/ee11/servlet/ServletApiResponse.java:
```diff
@@ -632,4 +656,132 @@ public String toString()
+        @Override
+        public void setContentLengthLong(long len)
+        {
+            // NOOP for include.
```
See above.

janbartel commented on jetty-ee11/jetty-ee11-servlet/src/main/java/org/eclipse/jetty/ee11/servlet/ServletApiResponse.java:
```diff
@@ -632,4 +656,132 @@ public String toString()
+        @Override
+        public void setContentType(String type)
+        {
+            // NOOP for include.
```
See above

janbartel commented on jetty-ee11/jetty-ee11-servlet/src/main/java/org/eclipse/jetty/ee11/servlet/ServletApiResponse.java:
```diff
@@ -632,4 +656,132 @@ public String toString()
+        @Override
+        public void reset()
+        {
+            // NOOP for include.
```
As above

janbartel commented on jetty-ee11/jetty-ee11-servlet/src/main/java/org/eclipse/jetty/ee11/servlet/ServletApiResponse.java:
```diff
@@ -632,4 +656,132 @@ public String toString()
+        }
+
+        @Override
+        public void resetBuffer()
```
Same comment for every method below, _except_ for the `sendRedirect` ones.

janbartel commented on jetty-ee11/jetty-ee11-servlet/src/main/java/org/eclipse/jetty/ee11/servlet/ServletApiResponse.java:
```diff
@@ -632,4 +656,132 @@ public String toString()
         }
     }
+
+    public static class CrossContextInclude extends ServletApiResponse
```
I think we need a comment about why this class is different from the `CrossContextDispatcher.IncludeResponse`

gregw committed:
Merge branch 'jetty-12.1.x' into jetty-12.1.0/redirect

gregw committed:
Update javadoc from review

Also update EE10 to also noop included response methods

gregw committed:
Update javadoc from review

Also update EE10 to also noop included response methods

gregw committed:
Update javadoc from review

Also update EE10 to also noop included response methods

gregw committed:
Merge branch 'jetty-12.1.x' into jetty-12.1.0/redirect

gregw commented:
@sbordet @janbartel @lachlan-roberts nudge!

janbartel commented on jetty-ee11/jetty-ee11-servlet/src/main/java/org/eclipse/jetty/ee11/servlet/ServletApiResponse.java:
```diff
@@ -632,4 +658,137 @@ public String toString()
     }
+
+    /**
+     * Servlet API wrapper for cross context included responses.
```
```suggestion
     * Servlet API wrapper used on the post-dispatch side of a cross-context include.
```

janbartel reviewed (approved):
Made a suggestion to clarify exactly how the `CrossContextInclude` wrapper is used. Other than that, looks fine, so +1

gregw commented:
The CI failure is due to the EE1 JSP artifact not being released yet.  Builds locally OK.

gregw committed:
Update jetty-ee11/jetty-ee11-servlet/src/main/java/org/eclipse/jetty/ee11/servlet/ServletApiResponse.java

Co-authored-by: Jan Bartel <janb@webtide.com>

joakime commented:
> The CI failure is due to the EE1 JSP artifact not being released yet. Builds locally OK.

I went ahead and released 11.0.0-M19 of the mortbay jsp-impl jars.
And updated this branch to use them.

* https://github.com/jetty-project/jasper-jsp/releases/tag/jasper-jsp-11.0.0-M19
* https://repo1.maven.org/maven2/org/mortbay/jasper/apache-jsp/11.0.0-M19/

joakime commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Response.java:
```diff
@@ -283,23 +283,47 @@ static void sendRedirect(Request request, Response response, Callback callback,
+        if (!HttpStatus.isRedirection(code))
         {
-            callback.failed(new IllegalArgumentException("No location"));
+            callback.failed(new IllegalArgumentException("Not a 3xx redirect code"));
```
The change to `isRedirection(int)` to not consider 304 a redirect means that this IAE can happen on a 304 now?

Of course, that would mean someone using sendRedirect with a 304 status code.  Which is just a strange thing to do.

sbordet replied:
Considering also the check for `location` below, seems that there really is a class of "redirect to location" codes that is worth a method in `HttpStatus`.
And yes, an application calling this method with 304 should get a ISE.

joakime commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpStatus.java:
```diff
@@ -383,7 +383,7 @@ public static boolean isSuccess(int code)
     public static boolean isRedirection(int code)
     {
-        return ((300 <= code) && (code <= 399));
+        return ((300 <= code) && (code <= 399) && code != 304);
```
Seems like you added this extra 304 status code hack here just to avoid producing a response body.
Seems like the wrong place for this special case.

Also, with 304 we should never produce trailers.
Perhaps this needs to be somewhere else?

sbordet replied:
Strictly per RFC, 304 is a redirection status code.

We have a similar issue for informational and interim status codes, and ended up having 2 methods.
Perhaps add a `isRedirect(int)` or `isLocationRedirect(int)` and exclude 304 in this new method?

gregw replied:
@joakime No.  This is to prevent 304 from being used with a sendRedirect.  But 304 is a redirect, so I will do it elsewhere.

joakime commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Response.java:
```diff
@@ -317,9 +341,22 @@ static void sendRedirect(Request request, Response response, Callback callback,
         }
 
-        response.getHeaders().put(HttpHeader.LOCATION, toRedirectURI(request, location));
+        if (content == null && request.getConnectionMetaData().getHttpConfiguration().isGenerateRedirectBody())
```
If `isGenerateRedirectBody()` is false, then the trailers for this response should also not be produced.

gregw replied:
We can still to trailers.  This is a different "no body" that for responses that have no body. Rather it is saying "no body" as in we do not generate a HTML redirection message (as we never have done), but it is still a message that can have a body, so trailers can still be sent.    The "no body at all" handling is done in response generation

joakime reviewed (dismissed):
The 304 status code special case seems to be in the wrong place.
Also the other parts of the spec related to 3xx and response bodies say that if there's no response body, there is also no trailers produced.

joakime committed:
Updating ee11 jsp-impl to 11.0.0-M19

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ConnectionMetaData.java:
```diff
@@ -52,7 +52,9 @@ default boolean isSecure()
      * @return whether the functionality of pushing resources is supported
+     * @deprecated in favour of 103 Early Hints
      */
+    @Deprecated
```
Orthogonal change?

gregw replied:
yes sorry.  We have not been reviewing 12.1 changes unless necessary. I thought the redirect changes needed review, but some other servlet 6.1 changes got swept up.

sbordet commented on jetty-ee11/jetty-ee11-servlet/src/main/java/org/eclipse/jetty/ee11/servlet/PushBuilderImpl.java:
```diff
@@ -23,6 +23,7 @@
 import org.eclipse.jetty.http.MetaData;
 import org.eclipse.jetty.util.URIUtil;
 
+@Deprecated(forRemoval = true)
```
Orthogonal change?
Also, missing `since` attribute?

gregw replied:
yeah, this started off as a general 6.1 branch update.  We are mostly doing that without PRs, but the other changes needed review

sbordet commented on jetty-ee11/jetty-ee11-servlet/src/main/java/org/eclipse/jetty/ee11/servlet/ServletApiRequest.java:
```diff
@@ -739,6 +739,7 @@ public <T extends HttpUpgradeHandler> T upgrade(Class<T> handlerClass) throws IO
     }
 
     @Override
+    @Deprecated(forRemoval = true)
```
Ditto

gregw committed:
Updates from review

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpStatus.java:
```diff
@@ -246,7 +246,7 @@ public boolean isSuccess()
         public boolean isRedirection()
         {
-            return HttpStatus.isRedirection(this._code);
+            return HttpStatus.isRedirectWithLocation(this._code);
```
I think this is possibly a breaking behavior, so I would leave it unchanged.

sbordet commented on jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/HttpStatus.java:
```diff
@@ -381,11 +381,27 @@ public static boolean isSuccess(int code)
      * <code>Redirection</code> messages.
      */
-    public static boolean isRedirection(int code)
+    public static boolean isRedirect(int code)
```
The RFC naming is actually "redirection", as also the javadoc of this method uses.

I would keep this method named `isRedirection()` with 300-399, and just add `isLocationRedirect()` or `isRedirectWithLocation()` (I prefer the former).

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ConnectionMetaData.java:
```diff
@@ -52,7 +52,9 @@ default boolean isSecure()
      * @return whether the functionality of pushing resources is supported
+     * @deprecated in favour of 103 Early Hints
      */
+    @Deprecated(since = "12.0.1")
```
Remove this, as I understand from your comment was slipped in unintentionally?

joakime replied:
It is also the wrong version string

sbordet commented on jetty-ee11/jetty-ee11-servlet/src/main/java/org/eclipse/jetty/ee11/servlet/PushBuilderImpl.java:
```diff
@@ -23,6 +23,7 @@
 import org.eclipse.jetty.http.MetaData;
 import org.eclipse.jetty.util.URIUtil;
 
+@Deprecated(forRemoval = true, since = "12.1.0")
```
Ditto remove?

gregw replied:
OK I'll remove here. I'll just directly commit to the branch as we are doing with other 6.1 changes.  I am only PRing this change because it affects core.

sbordet commented on jetty-ee11/jetty-ee11-servlet/src/main/java/org/eclipse/jetty/ee11/servlet/ServletApiRequest.java:
```diff
@@ -739,6 +739,7 @@ public <T extends HttpUpgradeHandler> T upgrade(Class<T> handlerClass) throws IO
     }
 
     @Override
+    @Deprecated(forRemoval = true, since = "12.1.0")
```
Ditto remove?

gregw committed:
Merge branch 'jetty-12.1.x' into jetty-12.1.0/redirect

gregw committed:
updates from review

gregw committed:
Merge branch 'jetty-12.1.x' into jetty-12.1.0/redirect

gregw commented:
@sbordet nudge

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Response.java:
```diff
@@ -317,9 +341,22 @@ static void sendRedirect(Request request, Response response, Callback callback,
+            String body = """
+            <!DOCTYPE html>
+            <html lang="en">
+            <head><meta charset="ISO-8859-1"/><meta http-equiv="refresh" content="0; URL=%s/"/><title>Redirecting...</title></head>
```
There is an extra "/":

```suggestion
            <head><meta charset="ISO-8859-1"/><meta http-equiv="refresh" content="0; URL=%s"/><title>Redirecting...</title></head>
```

gregw committed:
Update jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Response.java

Co-authored-by: Simone Bordet <simone.bordet@gmail.com>
